File: leetcode/chain/2_2.cpp
#include <iostream>

#include "list_node.h"

class Solution {
 public:
  ListNode* kth_to_last(ListNode* node, int k) {
    ListNode *slow = node, *fast = node;
    for (int i = 0; i < k; ++i) {
      fast = fast->next;
    }
    while (fast != nullptr) {
      fast = fast->next;
      slow = slow->next;
    }
    return slow;
  }
};

int main() {
  ListNode* node = new ListNode(1);
  node->append(new ListNode(2));
  node->append(new ListNode(4));
  node->append(new ListNode(3));
  node->append(new ListNode(4));
  node->append(new ListNode(5));
  node->print();

  Solution solution = Solution();
  ListNode* ret = solution.kth_to_last(node, 4);
  std::cout << ret->val << std::endl;
}

File: leetcode/chain/2_2.py
# Definition for singly-linked list.
class ListNode:

    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:

    def kthToLast(self, head: ListNode, k: int) -> int:
        '''返回倒数第 k 个节点
        
        @Note:
            p,q都指向head,p先走k步,然后pq都往前走,直到p为None
            快慢指针
        '''
        p, q = head, head
        for i in range(k):
            p = p.next
        while p is not None:
            p = p.next
            q = q.next
        return q.val
